ferrocene/tools/self-test/src/linkers.rs

```diff
@@ -99,6 +187,249 @@ pub(crate) fn check_and_add_rustflags(
     Ok(())
 }
 
+/// Look at the arguments given to the linker.
+///
+/// * Returns `Ok(true)` if these linker arguments look OK
+/// * Returns `Ok(false)` if the linker arguments were not OK, but an extra
+///   argument was added to `compiler_args` and so the caller should try again
+/// * Returns `Err(...)` if the linker arguments were not OK, and there's no
+///   proposed remedy
+fn linker_args_ok<'a, I>(
+    target: &str,
+    linker_args: I,
+    compiler_args: &mut Vec<String>,
+) -> Result<bool, Error>
+where
+    I: Iterator<Item = &'a str>,
+{
+    let linker_args = rationalise_linker_args(linker_args);
+    for arg in linker_args {
+        match arg {
+            LinkerArg::Input(_) => {}
+            LinkerArg::Output(_) => {}
+            LinkerArg::LibraryPath(_) => {}
+            LinkerArg::DiscardAll => {}
+            LinkerArg::Keyword(_) => {}
+            LinkerArg::Link(_) => {}
+            LinkerArg::Emulation(_) => {}
+            LinkerArg::PluginOpt(_) => {}
+            LinkerArg::LittleEndian => {}
+            LinkerArg::PicExecutable => {}
+            LinkerArg::NonPicExecutable => {}
+            LinkerArg::DynamicLinker(_) => {}
+            LinkerArg::Sysroot(_) => {}
+            LinkerArg::BuildId => {}
+            LinkerArg::EhFrameHeader => {}
+            LinkerArg::HashStyle(_) => {}
+            LinkerArg::AsNeeded => {}
+            LinkerArg::NoAsNeeded => {}
+            LinkerArg::PushState => {}
+            LinkerArg::PopState => {}
+            LinkerArg::FixCortexA53_843419 => {}
```

Comment:
https://reviews.llvm.org/D36749

Reply:
> This patch provides the mechanism to fix instances of the instruction sequence that may trigger the cortex-a53 843419 erratum. The fix is provided by replacing one of the instructions in the erratum sequence with a branch to a replacement patch sequence. The patch sequence is just the same instruction, followed by a branch back to the instruction following the one we replaced. The act of adding the branch is sufficient to prevent the erratum.

Reply:
https://developer.arm.com/documentation/epm048406/latest

> When executing in AArch64 state, a load or store instruction which uses the result of an ADRP instruction as a base register, or which uses a base register written by an instruction immediately after an ADRP to the same register, might access an incorrect address.

I only found it because I saw GCC emit it. I don't know how we would go about finding any other important errata fixes that should be opted in to, without reading the ld source code in detail. It sounds like something a FuSa client should be talking to Arm about, getting a recommendation, and then adding it to their `-Clink-args=xxx` configuration.

Reply:
On the Linux kernel, the flag is opt-in: https://patchwork.kernel.org/project/linux-arm-kernel/patch/1471873737-7614-1-git-send-email-will.deacon@arm.com/

Reply:
> On the Linux kernel, the flag is opt-in

Hmm, it seems like the kernel flag is [enabled by default](https://github.com/torvalds/linux/blob/c42d9eeef8e5ba9292eda36fd8e3c11f35ee065c/arch/arm64/Kconfig#L607).

> I only found it because I saw GCC emit it. I don't know how we would go about finding any other important errata fixes that should be opted in to, without reading the ld source code in detail.

Same :) But we are aware of it now, and we should make sure we address it IMO. I'll open a separate issue about it.

> It sounds like something a FuSa client should be talking to Arm about, getting a recommendation, and then adding it to their -Clink-args=xxx configuration.

Customers cannot add custom `-Clink-args=xxx` on their own without invalidating Ferrocene's qualification, unless it's a flag we explicitly list (and test): [(1)](https://public-docs.ferrocene.dev/main/safety-manual/options.html) [(2)](https://public-docs.ferrocene.dev/main/user-manual/cli.html#um_rustc_C_link_args).

Reply:
Opened https://github.com/ferrocene/ferrocene/issues/125 for followup, we can handle it in a separate PR.